charts/Store.ts

```diff
@@ -0,0 +1,115 @@
+import { Indicator } from "./Indicator"
+import { observable, runInAction, computed } from "mobx"
+import { fetchJSON, difference, values } from "./Util"
+import { BAKED_BASE_URL } from "settings"
+
+export class StoreEntry<EntityType> {
+    @observable.ref isLoading: boolean = false
+    @observable.ref lastRetrieved?: Date = undefined
+    @observable.ref error?: Error = undefined
+    @observable.ref entity?: EntityType
+
+    constructor(entity?: EntityType) {
+        if (entity) this.entity = entity
+    }
+}
+
+export interface StoreEntries<EntityType> {
+    [id: number]: StoreEntry<EntityType>
+}
+
+export class IndicatorStore {
+    @observable indicatorsById: StoreEntries<Indicator> = {}
+    fetchAllPromise?: Promise<Indicator[]>
+
+    private async fetchAll(): Promise<Indicator[]> {
+        try {
+            const json = await fetchJSON(
+                `${BAKED_BASE_URL}/explore/indicators.json`
+            )
+            const indicators: Indicator[] = json.indicators
+            return indicators
+        } catch (error) {
+            // TODO bubble up error to UI
+            console.error("Failed to fetch indicators", error)
+            return []
+        }
+    }
+
+    private loadAllIdempotent(): Promise<Indicator[]> {
+        if (!this.fetchAllPromise) {
+            this.fetchAllPromise = new Promise(async resolve => {
+                const indicators = await this.fetchAll()
+
+                // Find all IDs which are marked as loading but were not
+                // retrieved in fetchAll()
+                const loadingIds = Object.entries(this.indicatorsById)
+                    .filter(([id, entry]) => entry.isLoading)
+                    .map(([id]) => parseInt(id))
+                const loadedIds = indicators.map(i => i.id)
+                const missingIds = difference(loadingIds, loadedIds)
+
+                runInAction(() => {
+                    // Update each loaded indicator in place
+                    indicators.forEach(indicator => {
+                        const storeEntry = this.get(indicator.id)
+                        storeEntry.isLoading = false
+                        storeEntry.lastRetrieved = new Date()
+                        storeEntry.error = undefined
+                        storeEntry.entity = indicator
+                    })
+                    // Mark all other indicator as failed
+                    missingIds.forEach(id => {
+                        const storeEntry = this.get(id)
+                        storeEntry.isLoading = false
+                        storeEntry.lastRetrieved = undefined
+                        storeEntry.error = new Error(
+                            `No indicator with id ${id}`
+                        )
+                    })
+                })
```

Comment:
I don't really understand why all this is needed. When are we ever going to have missing indicators?

Reply:
I think when we delete a chart or we remove its "explorable" status. The user would arrive on the page, a placeholder indicator will be created, but the indicator won't exist in `indicators.json`.

Reply:
Ah, is the concern about stale links basically? With old IDs? That makes sense